Add option to exclude Realm file from iCloud backup

danibonilha opened the pull request:
## What, How & Why?
This PR adds a flag on Realm Config to exclude realm files from iCloud backup and also removes a reference from a guide that has been deleted.

This closes #4139

## Draft discussion
@kraenhansen I'm starting this as a draft to make sure I'm on the right path, and to align some things:

1.  In order to add this flag, some changes are required on @realm/core, to basically add `exclude_from_icloud_backup` on `spec.yml` and `shared_realm.hpp`, how is that coordinated ?

2.  You mentioned testing might be tricky for this one, it's a bit concerning if it's too difficult because I might not have the bandwidth to do it, but I can give it a try if there are some really similar examples.

3. This implementation is enough to ENABLE excluding from backup, but it cannot be DISABLED once a realm file has been set with  `excludedFromBackup`,  since we are just checking for `true` on the flag and then setting the `NSURLIsExcludedFromBackupKey` key, there's no check to remove this key when the flag is false again. Which means that, if a realm file has been created with `excludeFromIcloudBackup=false` and then set to `excludeFromIcloudBackup=true` the file will be correctly excluded, but if the same file if set again to `excludeFromIcloudBackup=false` it will remain excluded. I see two ways to handle this:   

 - 1. Stick with this implementation and explain the behavior via docs;
 - 2. Always check flag status and update files accordingly; 

- Particularly I don't think there's much usage of setting this flag dynamically, so i'd probably go with the first option since realm does provide an easy way to copy a file with different configs if needed, although the second one is more flexible, it might take longer to be done as I believe I'd need to always be checking if the files already have the excludedFromBackup key set. 

4. About the .md file on how to test this fix,  where would that live?

## Testing the fix
Generate the build files and test it on a sample react-native app on an IOS simulator, adding `excludeFromIcloudBackup: true` to the Realm configuration. 

To validate if the file has the exclude backup attribute, you need to check the files attributes with `xattr` on your terminal, in order to do so you need the path of realm files from the simulator Documents' folder, you can easily get that first by running:
```shell
open `xcrun simctl get_app_container booted com.your.app.bundleId data`/Documents
```
This will open a Finder window with the files, you can just drag and drop it to the terminal after adding `xattr`
 ```shell
 xattr [realm file simulator path]
 ```
 if this command returns:
 ```shell
 com.apple.metadata:com_apple_backup_excludeItem
 ```
It means the file has been successfully marked as excluded from backup 🎉 
if it returns nothing, it means the file has no attributes therefore is not excluded from backup. 

## ☑️ ToDos
<!-- Add your own todos here -->
* [ ] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [ ] 📝 Update `COMPATIBILITY.md`
* [ ] 🚦 Tests
* [ ] 📦 Updated internal package version in consuming `package.json`s (if updating internal packages)
* [x] 📱 Check the React Native/other sample apps work if necessary
* [x] 💥 `Breaking` label has been applied or is not necessary
* [ ] 🔔 Mention `@realm/devdocs` if documentation changes are needed

kraenhansen commented on packages/realm/bindgen/js_opt_in_spec.yml:
```diff
@@ -51,6 +51,7 @@ records:
       - initialization_function
       - should_compact_on_launch_function
       - automatically_handle_backlinks_in_migrations
+      - exclude_from_icloud_backup
```
These values refer to fields of the `RealmConfig` defined by Realm Core: https://github.com/realm/realm-core/blob/43a1aac3dc41edf80a5be093a8a65e839b0ad225/src/realm/object-store/shared_realm.hpp#L94-L194

If it's important for this behavior to be configurable, an alternative that wouldn't require changes to Core would be pass this as a bool argument to `after_realm_open` 🤔 Or rename `after_realm_open` to `exclude_from_icloud_backup` and call it conditionally in `Realm.ts`.

danibonilha replied:
Oh at first I thought the only way to add parameters to the config was adding on realm-core, but I see now we can scope it just to realm-js, so I'll go with the parameter option.

kraenhansen commented on packages/realm/bindgen/js_spec.yml:
```diff
@@ -15,6 +15,7 @@ classes:
       remove_file: '(path: const std::string&)'
       remove_directory: '(path: const std::string&)'
       get_cpu_arch: () -> std::string
+      after_realm_open: '(sharedRealm: std::shared_ptr<Realm>)'
```
I suggest naming this `realm` as all references to `Realm` classes in the codebase are to objects wrapped in `shared_ptr`. We also refer to these are `SharedRealm` (made possible by [this line in Realm Core's spec.yml](https://github.com/realm/realm-core/blob/85eeca41654cc9070ad81a21a4b1d153ad467c33/bindgen/spec.yml#L928-L929))

```suggestion
      after_realm_open: '(realm: SharedRealm)'
```

kraenhansen commented on packages/realm/binding/apple/platform.mm:
```diff
@@ -158,6 +163,18 @@
 
+void JsPlatformHelpers::after_realm_open(const std::shared_ptr<Realm> sharedRealm) {
+     if(sharedRealm->config().exclude_from_icloud_backup) {
+       RLMAddSkipBackupAttributeToItemAtPath(sharedRealm->config().path);
```
Nit: It could make sense to introduce a local `const auto path = sharedRealm->config().path` to keep things DRY.

kraenhansen commented on packages/realm/binding/platform.hpp:
```diff
@@ -19,6 +19,7 @@
 #pragma once
 
 #include <string>
+#include "realm/object-store/shared_realm.hpp"
```
As this is referring to an include outside of the JS SDK code-base:

```suggestion
#include <realm/object-store/shared_realm.hpp>
```

kraenhansen commented:
> In order to add this flag, some changes are required on @realm/core, to basically add `exclude_from_icloud_backup` on `spec.yml` and `shared_realm.hpp`, how is that coordinated?

This can be done through a PR to https://github.com/realm/realm-core and and updating to the submodule of this repo to point to the PR commit in realm-core. But, since this won't be generally supported by the other SDKs I would rather that we keep changes to this codebase. A possible alternative is suggested in this comment: https://github.com/realm/realm-js/pull/6922#discussion_r1832256555. If you _really_ want this to be supported generally across SDKs, I'd say the logic calling into `setResourceValue` should live in Realm Core too.

> Always check flag status and update files accordingly

I like this idea 👍 I agree this would make it more flexible.

> it might take longer to be done as I believe I'd need to always be checking if the files already have the excludedFromBackup key set.

Would it throw if you set it when it's already set?

kraenhansen commented on packages/realm/binding/apple/platform.mm:
```diff
@@ -18,6 +18,7 @@
 
 #include "../platform.hpp"
 
+#include <memory>
```
Is this strictly needed?

danibonilha replied:
I don't think so, it's probably just a leftover

danibonilha commented:
> But, since this won't be generally supported by the other SDKs I would rather that we keep changes to this codebase.

Agreed, I've added a new implementation with a parameter 

>Would it throw if you set it when it's already set? If not, I wouldn't be too worried of setting it while it's already set, as we'd expect it to be called once per app open.

It does not throw, so I went ahead and update it to dynamically change based on the attribute

>Perhaps a guide-testing-exclude-icloud-backup.md

I've created this file, but also added a script to make it easier to test, I have created for me but I just figured it would be helpful to have it in this project

Oh I also updated the `ninja` requirement, as I was having trouble running for android and it was because this package was missing, so I added a step with the OpenJDK.

kraenhansen commented:
Awesome! I assume you've already verified this fixes the issue in your own (or some test) app?

danibonilha commented:
> Awesome! I assume you've already verified this fixes the issue in your own (or some test) app?

Yes I have, it's working perfectly on my app: 

`excludeFromIcloudBackup: true`

<img width="554" alt="image" src="https://github.com/user-attachments/assets/957f6661-a504-42cd-b2da-ce13b6a85a2b">


`excludeFromIcloudBackup: false `

<img width="329" alt="image" src="https://github.com/user-attachments/assets/969af484-97f8-470e-90e5-9d2fc238d90b">

I'm having issues though running the android app, but it might be my environment or something, is there a check on the CI that builds the android sample app ? Since there's no implementation there, just having it the app building should be enough.

kraenhansen commented:
I reworded the note in the changelog and I ran the tests locally on Android, iOS and Node.js 👍 I'll go ahead with the merge and release of this 👍

kraenhansen commented:
I'm seeing some compilation errors on Node.js for Windows that I need to resolve before releasing this 😕 Stay tuned.

danibonilha commented:
> I'm seeing some compilation errors on Node.js for Windows that I need to resolve before releasing this 😕 Stay tuned.

Could it be related maybe to using unnamed parameters on platform.cpp ?

kraenhansen commented:
Sorry about the wait on this one. Would you mind taking a look and providing a review on https://github.com/realm/realm-js/pull/6927?
